File: sdk/js/src/profiles.ts
/**
 * Model profile management for FlatAgents.
 *
 * Provides centralized model configuration through profiles.yml files.
 * Profiles enable easy switching between model configurations (e.g., dev vs prod).
 *
 * Resolution order (low to high priority):
 * 1. default profile (fallback)
 * 2. Named profile from agent's model field
 * 3. Inline overrides from agent's model field
 * 4. override profile (trumps all)
 */

import { readFileSync, existsSync } from "fs";
import { join, dirname } from "path";
import yaml from "yaml";
import type { ModelConfig, ModelProfileConfig, ProfilesConfig } from "./types";

// Cache loaded profile managers by directory
const profileManagerCache: Map<string, ProfileManager> = new Map();

/**
 * Manages model profiles from profiles.yml files.
 *
 * Resolution order (low to high priority):
 * 1. default profile (fallback)
 * 2. Named profile from agent's model field
 * 3. Inline overrides from agent's model field
 * 4. override profile (trumps all)
 *
 * @example
 * ```typescript
 * const manager = new ProfileManager("config/profiles.yml");
 * const config = manager.resolveModelConfig("fast-cheap");
 * console.log(config);
 * // { provider: 'cerebras', name: 'zai-glm-4.6', temperature: 0.6 }
 * ```
 */
export class ProfileManager {
  private profiles: Record<string, ModelProfileConfig> = {};
  private defaultProfile: string | undefined;
  private overrideProfile: string | undefined;
  readonly profilesFile: string | undefined;

  constructor(profilesFile?: string) {
    this.profilesFile = profilesFile;
    if (profilesFile) {
      this.loadProfiles(profilesFile);
    }
  }

  /**
   * Get or create a ProfileManager for a directory.
   * Caches instances by directory to avoid re-reading profiles.yml.
   */
  static getInstance(configDir: string): ProfileManager {
    if (!profileManagerCache.has(configDir)) {
      const profilesPath = join(configDir, "profiles.yml");
      if (existsSync(profilesPath)) {
        profileManagerCache.set(configDir, new ProfileManager(profilesPath));
      } else {
        // No profiles file - return empty manager
        profileManagerCache.set(configDir, new ProfileManager());
      }
    }
    return profileManagerCache.get(configDir)!;
  }

  /**
   * Clear cached ProfileManager instances.
   */
  static clearCache(): void {
    profileManagerCache.clear();
  }

  private loadProfiles(profilesFile: string): void {
    if (!existsSync(profilesFile)) {
      return;
    }

    const content = readFileSync(profilesFile, "utf-8");
    const config = yaml.parse(content) as ProfilesConfig | ProfilesConfig["data"];

    // Validate spec if present
    if ("spec" in config && config.spec !== "flatprofiles") {
      throw new Error(
        `Invalid profiles spec: expected 'flatprofiles', got '${config.spec}'`
      );
    }

    // Support both wrapped (spec/data) and unwrapped format
    const data = "data" in config ? config.data : config;

    this.profiles = data.model_profiles ?? {};
    this.defaultProfile = data.default;
    this.overrideProfile = data.override;
  }

  /**
   * Get a profile by name.
   */
  getProfile(name: string): ModelProfileConfig | undefined {
    return this.profiles[name];
  }

  /**
   * Get all loaded profiles.
   */
  getProfiles(): Record<string, ModelProfileConfig> {
    return this.profiles;
  }

  /**
   * Get the default profile name.
   */
  getDefaultProfile(): string | undefined {
    return this.defaultProfile;
  }

  /**
   * Get the override profile name.
   */
  getOverrideProfile(): string | undefined {
    return this.overrideProfile;
  }

  /**
   * Resolve the final model configuration.
   *
   * Resolution order:
   * 1. Start with default profile (if set)
   * 2. Apply named profile (if agentModelConfig is string or has 'profile' key)
   * 3. Merge inline overrides (if agentModelConfig is dict)
   * 4. Apply override profile (trumps all)
   *
   * @param agentModelConfig - Agent's model config (string, object, or undefined)
   * @returns Fully resolved model configuration
   * @throws Error if a referenced profile is not found
   */
  resolveModelConfig(
    agentModelConfig: string | ModelConfig | undefined
  ): ModelConfig {
    const result: Partial<ModelConfig> = {};

    // 1. Apply default profile
    if (this.defaultProfile) {
      const defaultCfg = this.getProfile(this.defaultProfile);
      if (defaultCfg) {
        Object.assign(result, defaultCfg);
      } else {
        console.warn(`Default profile '${this.defaultProfile}' not found`);
      }
    }

    // 2. Handle agent's model config
    if (typeof agentModelConfig === "string") {
      // String = profile name
      const profileCfg = this.getProfile(agentModelConfig);
      if (profileCfg) {
        Object.assign(result, profileCfg);
      } else {
        throw new Error(`Model profile '${agentModelConfig}' not found`);
      }
    } else if (typeof agentModelConfig === "object" && agentModelConfig) {
      // Check for profile reference in object
      const profileName = (agentModelConfig as any).profile;
      if (profileName) {
        const profileCfg = this.getProfile(profileName);
        if (profileCfg) {
          Object.assign(result, profileCfg);
        } else {
          throw new Error(`Model profile '${profileName}' not found`);
        }
      }

      // Merge inline overrides (excluding 'profile' key)
      const { profile, ...inlineOverrides } = agentModelConfig as any;
      for (const [key, value] of Object.entries(inlineOverrides)) {
        if (value !== undefined && value !== null) {
          (result as any)[key] = value;
        }
      }
    }

    // 3. Apply override profile (trumps all)
    if (this.overrideProfile) {
      const overrideCfg = this.getProfile(this.overrideProfile);
      if (overrideCfg) {
        Object.assign(result, overrideCfg);
      } else {
        console.warn(`Override profile '${this.overrideProfile}' not found`);
      }
    }

    return result as ModelConfig;
  }
}

/**
 * Convenience function to resolve model configuration.
 *
 * @param agentModelConfig - Agent's model config (string, object, or undefined)
 * @param configDir - Directory containing agent config (for profiles.yml lookup)
 * @param profilesFile - Explicit path to profiles.yml (overrides auto-discovery)
 * @returns Fully resolved model configuration
 *
 * @example
 * ```typescript
 * const config = resolveModelConfig("fast-cheap", "./config");
 * console.log(config);
 * // { provider: 'cerebras', name: 'zai-glm-4.6', temperature: 0.6 }
 * ```
 */
export function resolveModelConfig(
  agentModelConfig: string | ModelConfig | undefined,
  configDir: string,
  profilesFile?: string
): ModelConfig {
  const manager = profilesFile
    ? new ProfileManager(profilesFile)
    : ProfileManager.getInstance(configDir);

  return manager.resolveModelConfig(agentModelConfig);
}

File: sdk/python/flatagents/profiles.py
"""
Model profile management for FlatAgents.

Provides centralized model configuration through profiles.yml files.
Profiles enable easy switching between model configurations (e.g., dev vs prod).

Resolution order (low to high priority):
1. default profile (fallback)
2. Named profile from agent's model field
3. Inline overrides from agent's model field
4. override profile (trumps all)
"""

import os
from typing import Any, Dict, Optional

from .monitoring import get_logger

logger = get_logger(__name__)

# Cache loaded profile managers by directory
_profile_managers: Dict[str, "ProfileManager"] = {}


class ProfileManager:
    """
    Manages model profiles from profiles.yml files.

    Resolution order (low to high priority):
    1. default profile (fallback)
    2. Named profile from agent's model field
    3. Inline overrides from agent's model field
    4. override profile (trumps all)

    Example:
        >>> manager = ProfileManager("config/profiles.yml")
        >>> config = manager.resolve_model_config("fast-cheap")
        >>> print(config)
        {'provider': 'cerebras', 'name': 'zai-glm-4.6', 'temperature': 0.6}
    """

    def __init__(self, profiles_file: Optional[str] = None):
        """
        Initialize profile manager.

        Args:
            profiles_file: Path to profiles.yml file (optional)
        """
        self.profiles_file = profiles_file
        self._profiles: Dict[str, Dict[str, Any]] = {}
        self._default_profile: Optional[str] = None
        self._override_profile: Optional[str] = None

        if profiles_file:
            self._load_profiles(profiles_file)

    @classmethod
    def get_instance(cls, config_dir: str) -> "ProfileManager":
        """
        Get or create a ProfileManager for a directory.

        Caches instances by directory to avoid re-reading profiles.yml.

        Args:
            config_dir: Directory containing profiles.yml

        Returns:
            ProfileManager instance (may have no profiles if file not found)
        """
        if config_dir not in _profile_managers:
            profiles_path = os.path.join(config_dir, "profiles.yml")
            if os.path.exists(profiles_path):
                _profile_managers[config_dir] = cls(profiles_path)
            else:
                # No profiles file - return empty manager
                _profile_managers[config_dir] = cls()
        return _profile_managers[config_dir]

    @classmethod
    def clear_cache(cls) -> None:
        """Clear cached ProfileManager instances."""
        _profile_managers.clear()

    def _load_profiles(self, profiles_file: str) -> None:
        """Load profiles from YAML file."""
        try:
            import yaml
        except ImportError:
            raise ImportError("pyyaml is required for profiles.yml")

        if not os.path.exists(profiles_file):
            logger.debug(f"No profiles file at {profiles_file}")
            return

        with open(profiles_file, 'r') as f:
            config = yaml.safe_load(f) or {}

        # Validate spec if present
        spec = config.get('spec')
        if spec and spec != 'flatprofiles':
            raise ValueError(f"Invalid profiles spec: expected 'flatprofiles', got '{spec}'")

        # Support both wrapped (spec/data) and unwrapped format
        data = config.get('data', config)

        self._profiles = data.get('model_profiles', {})
        self._default_profile = data.get('default')
        self._override_profile = data.get('override')

        logger.info(
            f"Loaded {len(self._profiles)} profiles from {profiles_file}"
            f" (default={self._default_profile}, override={self._override_profile})"
        )

    def get_profile(self, name: str) -> Optional[Dict[str, Any]]:
        """
        Get a profile by name.

        Args:
            name: Profile name

        Returns:
            Profile config dict or None if not found
        """
        return self._profiles.get(name)

    @property
    def profiles(self) -> Dict[str, Dict[str, Any]]:
        """All loaded profiles."""
        return self._profiles

    @property
    def default_profile(self) -> Optional[str]:
        """Name of the default profile."""
        return self._default_profile

    @property
    def override_profile(self) -> Optional[str]:
        """Name of the override profile."""
        return self._override_profile

    def resolve_model_config(
        self,
        agent_model_config: Any
    ) -> Dict[str, Any]:
        """
        Resolve the final model configuration.

        Resolution order:
        1. Start with default profile (if set)
        2. Apply named profile (if agent_model_config is string or has 'profile' key)
        3. Merge inline overrides (if agent_model_config is dict)
        4. Apply override profile (trumps all)

        Args:
            agent_model_config: Agent's model config (string, dict, or None)

        Returns:
            Fully resolved model configuration dict

        Raises:
            ValueError: If a referenced profile is not found
        """
        result = {}

        # 1. Apply default profile
        if self._default_profile:
            default_cfg = self.get_profile(self._default_profile)
            if default_cfg:
                result.update(default_cfg)
            else:
                logger.warning(f"Default profile '{self._default_profile}' not found")

        # 2. Handle agent's model config
        if isinstance(agent_model_config, str):
            # String = profile name
            profile_cfg = self.get_profile(agent_model_config)
            if profile_cfg:
                result.update(profile_cfg)
            else:
                raise ValueError(f"Model profile '{agent_model_config}' not found")

        elif isinstance(agent_model_config, dict):
            # Check for profile reference in dict
            profile_name = agent_model_config.get('profile')
            if profile_name:
                profile_cfg = self.get_profile(profile_name)
                if profile_cfg:
                    result.update(profile_cfg)
                else:
                    raise ValueError(f"Model profile '{profile_name}' not found")

            # Merge inline overrides (excluding 'profile' key)
            inline_overrides = {
                k: v for k, v in agent_model_config.items()
                if k != 'profile' and v is not None
            }
            result.update(inline_overrides)

        # 3. Apply override profile (trumps all)
        if self._override_profile:
            override_cfg = self.get_profile(self._override_profile)
            if override_cfg:
                result.update(override_cfg)
            else:
                logger.warning(f"Override profile '{self._override_profile}' not found")

        return result


def resolve_model_config(
    agent_model_config: Any,
    config_dir: str,
    profiles_file: Optional[str] = None
) -> Dict[str, Any]:
    """
    Convenience function to resolve model configuration.

    Args:
        agent_model_config: Agent's model config (string, dict, or None)
        config_dir: Directory containing agent config (for profiles.yml lookup)
        profiles_file: Explicit path to profiles.yml (overrides auto-discovery)

    Returns:
        Fully resolved model configuration dict

    Example:
        >>> config = resolve_model_config("fast-cheap", "./config")
        >>> print(config)
        {'provider': 'cerebras', 'name': 'zai-glm-4.6', 'temperature': 0.6}
    """
    if profiles_file:
        manager = ProfileManager(profiles_file)
    else:
        manager = ProfileManager.get_instance(config_dir)

    return manager.resolve_model_config(agent_model_config)


def discover_profiles_file(config_dir: str, explicit_path: Optional[str] = None) -> Optional[str]:
    """
    Discover profiles.yml in config_dir if not explicitly provided.

    This is used by FlatAgent and FlatMachine to auto-discover profiles.yml
    and propagate it to child agents/machines.

    Args:
        config_dir: Directory to search for profiles.yml
        explicit_path: Explicit path if already provided (returned as-is)

    Returns:
        Path to profiles.yml if found, explicit_path if provided, or None
    """
    if explicit_path:
        return explicit_path
    default_path = os.path.join(config_dir, 'profiles.yml')
    return default_path if os.path.exists(default_path) else None


__all__ = [
    "ProfileManager",
    "resolve_model_config",
    "discover_profiles_file",
]
